apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/bulk/ZstdGroupedWeight.scala

```diff
@@ -0,0 +1,123 @@
+
+import com.github.luben.zstd.ZstdDirectBufferCompressingStreamNoFinalizer
+import org.apache.pekko.stream.{Attributes, FlowShape, Inlet, Outlet}
+import org.apache.pekko.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import org.apache.pekko.util.ByteString
+
+import java.nio.ByteBuffer
+import java.util.concurrent.atomic.AtomicReference
+
+/** A Pekko GraphStage that zstd-compresses a stream of bytestrings, and splits the output into zstd objects of size (minWeight + delta).
+  * Somewhat similar to Pekko's built-in GroupedWeight, but outputs valid zstd compressed objects.
+  */
+case class ZstdGroupedWeight(minSize: Long) extends GraphStage[FlowShape[ByteString, ByteString]] {
+  require(minSize > 0, "minSize must be greater than 0")
+
+  val zstdTmpBufferSize = 10 * 1024 * 1024; // TODO(#3429): make configurable?
+
+  val in = Inlet[ByteString]("ZstdGroupedWeight.in")
+  val out = Outlet[ByteString]("ZstdGroupedWeight.out")
+  override val shape: FlowShape[ByteString, ByteString] = FlowShape(in, out)
+
+  override def initialAttributes: Attributes = Attributes.name("ZstdGroupedWeight")
+
+  private case class State(
+      bytes: ByteString,
+      left: Long,
+  ) {
+    def append(bs: ByteString): State =
+      State(
+        bytes ++ bs,
+        left - bs.size,
+      )
+
+  }
+  private object State {
+    def empty(): State = State(ByteString.empty, minSize)
+  }
+
+  class ZSTD(
+      val tmpBuffer: ByteBuffer,
+      val compressionLevel: Int = 3,
+  ) extends AutoCloseable {
+
+    val compressingStream =
+      new ZstdDirectBufferCompressingStreamNoFinalizer(tmpBuffer, compressionLevel)
+
+    def compress(input: ByteString): ByteString = {
+      val inputBB = ByteBuffer.allocateDirect(input.size)
+      inputBB.put(input.toArrayUnsafe())
+      inputBB.flip()
+      compressingStream.compress(inputBB)
+      compressingStream.flush()
+      tmpBuffer.flip()
+      val result = ByteString.fromByteBuffer(tmpBuffer)
+      tmpBuffer.clear()
+      result
+    }
+
+    def zstdFinish(): ByteString = {
+      tmpBuffer.flip()
+      val result = ByteString.fromByteBuffer(tmpBuffer)
+      tmpBuffer.clear()
+      compressingStream.close()
+      result
+    }
+
+    override def close(): Unit = {
+      compressingStream.close()
+    }
+  }
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      // TODO(#3429): consider implementing a pool of tmp buffers to avoid allocating a new one for each stage
+      private val tmpBuffer = ByteBuffer.allocateDirect(zstdTmpBufferSize)
+      private val zstd = new AtomicReference[ZSTD](new ZSTD(tmpBuffer, 3))
+      private val state: AtomicReference[State] = new AtomicReference[State](State.empty())
+
+      override def postStop(): Unit = {
+        super.postStop()
+        if (zstd.get() != null) {
+          zstd.get().close()
```

Comment:
Since you allocate the buffer in this graphStageLogic you also need to deref it here. (or move everything to Zstd)